// instagram_mqtt/dist/realtime/realtime.client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RealtimeClient = void 0;
const constants_1 = require("../constants");
const commands_1 = require("./commands");
const shared_1 = require("../shared");
const mqttot_1 = require("../mqttot");
const errors_1 = require("../errors");
const EventEmitter = require("eventemitter3");
const mixins_1 = require("./mixins");
class RealtimeClient extends EventEmitter {
    /**
     *
     * @param {IgApiClient} ig
     * @param mixins - by default MessageSync and Realtime mixins are used
     */
    constructor(ig, mixins = [new mixins_1.MessageSyncMixin(), new mixins_1.RealtimeSubMixin()]) {
        super();
        this.realtimeDebug = shared_1.debugChannel('realtime');
        this.messageDebug = this.realtimeDebug.extend('message');
        this.safeDisconnect = false;
        this.emitError = (e) => this.emit('error', e);
        this.emitWarning = (e) => this.emit('warning', e);
        this.ig = ig;
        this.realtimeDebug(`Applying mixins: ${mixins.map(m => m.name).join(', ')}`);
        mixins_1.applyMixins(mixins, this, this.ig);
    }
    get mqtt() {
        return this._mqtt;
    }
    setInitOptions(initOptions) {
        if (Array.isArray(initOptions))
            initOptions = { graphQlSubs: initOptions };
        this.initOptions = {
            graphQlSubs: [],
            skywalkerSubs: [],
            ...(initOptions || {}),
            socksOptions: typeof initOptions === 'object' && !Array.isArray(initOptions) ? initOptions.socksOptions : undefined,
        };
    }
    constructConnection() {
        const userAgent = this.ig.state.appUserAgent;
        const deviceId = this.ig.state.phoneId;
        const password = `sessionid=${this.ig.state.extractCookieValue('sessionid')}`;
        this.connection = new mqttot_1.MQTToTConnection({
            clientIdentifier: deviceId.substring(0, 20),
            clientInfo: {
                userId: BigInt(Number(this.ig.state.cookieUserId)),
                userAgent,
                clientCapabilities: 183,
                endpointCapabilities: 0,
                publishFormat: 1,
                noAutomaticForeground: false,
                makeUserAvailableInForeground: true,
                deviceId,
                isInitiallyForeground: true,
                networkType: 1,
                networkSubtype: 0,
                clientMqttSessionId: BigInt(Date.now()) & BigInt(0xffffffff),
                subscribeTopics: [88, 135, 149, 150, 133, 146],
                clientType: 'cookie_auth',
                appId: BigInt(567067343352427),
                deviceSecret: '',
                clientStack: 3,
                ...(this.initOptions.connectOverrides || {}),
            },
            password,
            appSpecificInfo: {
                app_version: this.ig.state.appVersion,
                'X-IG-Capabilities': this.ig.state.capabilitiesHeader,
                everclear_subscriptions: JSON.stringify({
                    inapp_notification_subscribe_comment: '17899377895239777',
                    inapp_notification_subscribe_comment_mention_and_reply: '17899377895239777',
                    video_call_participant_state_delivery: '17977239895057311',
                    presence_subscribe: '17846944882223835',
                }),
                'User-Agent': userAgent,
                'Accept-Language': this.ig.state.language.replace('_', '-'),
                platform: 'android',
                ig_mqtt_route: 'django',
                pubsub_msg_type_blacklist: 'direct, typing_type',
                auth_cache_enabled: '0',
            },
        });
    }
    async connect(initOptions) {
        var _a;
        this.realtimeDebug('Connecting to realtime-broker...');
        this.setInitOptions(initOptions);
        this.realtimeDebug(`Overriding: ${Object.keys(this.initOptions.connectOverrides || {}).join(', ')}`);
        this._mqtt = new mqttot_1.MQTToTClient({
            url: constants_1.REALTIME.HOST_NAME_V6,
            payloadProvider: () => {
                this.constructConnection();
                return shared_1.compressDeflate(this.connection.toThrift());
            },
            enableTrace: this.initOptions.enableTrace,
            autoReconnect: (_a = this.initOptions.autoReconnect) !== null && _a !== void 0 ? _a : true,
            requirePayload: false,
            socksOptions: this.initOptions.socksOptions,
            additionalOptions: this.initOptions.additionalTlsOptions,
        });
        this.commands = new commands_1.Commands(this.mqtt);
        this.direct = new commands_1.DirectCommands(this.mqtt);
        this.mqtt.on('message', async (msg) => {
            var _a;
            const unzipped = await shared_1.tryUnzipAsync(msg.payload);
            const topic = constants_1.RealtimeTopicsArray.find(t => t.id === msg.topic);
            if (topic && topic.parser && !topic.noParse) {
                const parsedMessages = topic.parser.parseMessage(topic, unzipped);
                this.messageDebug(`Received on ${topic.path}: ${JSON.stringify(Array.isArray(parsedMessages)
                    ? parsedMessages.map((x) => x.data)
                    : parsedMessages.data)}`);
                this.emit('receive', topic, Array.isArray(parsedMessages) ? parsedMessages : [parsedMessages]);
            }
            else {
                this.messageDebug(`Received raw on ${(_a = topic === null || topic === void 0 ? void 0 : topic.path) !== null && _a !== void 0 ? _a : msg.topic}: (${unzipped.byteLength} bytes) ${shared_1.prepareLogString(unzipped.toString())}`);
                this.emit('receiveRaw', msg);
            }
        });
        this.mqtt.on('error', e => this.emitError(e));
        this.mqtt.on('warning', w => this.emitWarning(w));
        this.mqtt.on('disconnect', () => this.safeDisconnect
            ? this.emit('disconnect')
            : this.emitError(new errors_1.ClientDisconnectedError('MQTToTClient got disconnected.')));
        return new Promise((resolve, reject) => {
            this.mqtt.on('connect', async () => {
                this.realtimeDebug('Connected. Checking initial subs.');
                const { graphQlSubs, skywalkerSubs, irisData } = this.initOptions;
                await Promise.all([
                    graphQlSubs && graphQlSubs.length > 0 ? this.graphQlSubscribe(graphQlSubs) : null,
                    skywalkerSubs && skywalkerSubs.length > 0 ? this.skywalkerSubscribe(skywalkerSubs) : null,
                    irisData ? this.irisSubscribe(irisData) : null,
                ]).then(resolve);
            });
            this.mqtt
                .connect({
                keepAlive: 20,
                protocolLevel: 3,
                clean: true,
                connectDelay: 60 * 1000,
            })
                .catch(e => {
                this.emitError(e);
                reject(e);
            });
        });
    }
    disconnect() {
        this.safeDisconnect = true;
        return this.mqtt.disconnect();
    }
    graphQlSubscribe(sub) {
        sub = typeof sub === 'string' ? [sub] : sub;
        this.realtimeDebug(`Subscribing with GraphQL to ${sub.join(', ')}`);
        return this.commands.updateSubscriptions({
            topic: constants_1.Topics.REALTIME_SUB,
            data: {
                sub,
            },
        });
    }
    skywalkerSubscribe(sub) {
        sub = typeof sub === 'string' ? [sub] : sub;
        this.realtimeDebug(`Subscribing with Skywalker to ${sub.join(', ')}`);
        return this.commands.updateSubscriptions({
            topic: constants_1.Topics.PUBSUB,
            data: {
                sub,
            },
        });
    }
    irisSubscribe({ seq_id, snapshot_at_ms, }) {
        this.realtimeDebug(`Iris Sub to: seqId: ${seq_id}, snapshot: ${snapshot_at_ms}`);
        return this.commands.updateSubscriptions({
            topic: constants_1.Topics.IRIS_SUB,
            data: {
                seq_id,
                snapshot_at_ms,
                snapshot_app_version: this.ig.state.appVersion
            },
        });
    }
}
exports.RealtimeClient = RealtimeClient;
