// mqtts/dist/mqtt.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.packetTypeToString = exports.PacketType = void 0;
var PacketType;
(function (PacketType) {
    PacketType[PacketType["Connect"] = 1] = "Connect";
    PacketType[PacketType["ConnAck"] = 2] = "ConnAck";
    PacketType[PacketType["Publish"] = 3] = "Publish";
    PacketType[PacketType["PubAck"] = 4] = "PubAck";
    PacketType[PacketType["PubRec"] = 5] = "PubRec";
    PacketType[PacketType["PubRel"] = 6] = "PubRel";
    PacketType[PacketType["PubComp"] = 7] = "PubComp";
    PacketType[PacketType["Subscribe"] = 8] = "Subscribe";
    PacketType[PacketType["SubAck"] = 9] = "SubAck";
    PacketType[PacketType["Unsubscribe"] = 10] = "Unsubscribe";
    PacketType[PacketType["UnsubAck"] = 11] = "UnsubAck";
    PacketType[PacketType["PingReq"] = 12] = "PingReq";
    PacketType[PacketType["PingResp"] = 13] = "PingResp";
    PacketType[PacketType["Disconnect"] = 14] = "Disconnect";
})(PacketType = exports.PacketType || (exports.PacketType = {}));
const reverseMapping = Object.fromEntries(Object.entries(PacketType).map(([k, v]) => [v, k]));
function packetTypeToString(type) {
    return reverseMapping[type].toUpperCase();
}
exports.packetTypeToString = packetTypeToString;
