// instagram_mqtt/dist/realtime/messages/message-sync.message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageSyncMessageTypes = void 0;
var MessageSyncMessageTypes;
(function (MessageSyncMessageTypes) {
    MessageSyncMessageTypes["Deletion"] = "deletion";
    MessageSyncMessageTypes["Media"] = "media";
    MessageSyncMessageTypes["Text"] = "text";
    MessageSyncMessageTypes["Like"] = "like";
    MessageSyncMessageTypes["Hashtag"] = "hashtag";
    MessageSyncMessageTypes["Profile"] = "profile";
    MessageSyncMessageTypes["MediaShare"] = "media_share";
    MessageSyncMessageTypes["Location"] = "location";
    MessageSyncMessageTypes["ActionLog"] = "action_log";
    MessageSyncMessageTypes["Title"] = "title";
    MessageSyncMessageTypes["UserReaction"] = "user_reaction";
    MessageSyncMessageTypes["HistoryEdit"] = "history_edit";
    MessageSyncMessageTypes["ReactionLog"] = "reaction_log";
    MessageSyncMessageTypes["ReelShare"] = "reel_share";
    MessageSyncMessageTypes["DeprecatedChannel"] = "deprecated_channel";
    MessageSyncMessageTypes["Link"] = "link";
    MessageSyncMessageTypes["RavenMedia"] = "raven_media";
    MessageSyncMessageTypes["LiveVideoShare"] = "live_video_share";
    MessageSyncMessageTypes["Test"] = "test";
    MessageSyncMessageTypes["StoryShare"] = "story_share";
    MessageSyncMessageTypes["ReelReact"] = "reel_react";
    MessageSyncMessageTypes["LiveInviteGuest"] = "live_invite_guest";
    MessageSyncMessageTypes["LiveViewerInvite"] = "live_viewer_invite";
    MessageSyncMessageTypes["TypeMax"] = "type_max";
    MessageSyncMessageTypes["Placeholder"] = "placeholder";
    MessageSyncMessageTypes["Product"] = "product";
    MessageSyncMessageTypes["ProductShare"] = "product_share";
    MessageSyncMessageTypes["VideoCallEvent"] = "video_call_event";
    MessageSyncMessageTypes["PollVote"] = "poll_vote";
    MessageSyncMessageTypes["FelixShare"] = "felix_share";
    MessageSyncMessageTypes["AnimatedMedia"] = "animated_media";
    MessageSyncMessageTypes["CtaLink"] = "cta_link";
    MessageSyncMessageTypes["VoiceMedia"] = "voice_media";
    MessageSyncMessageTypes["StaticSticker"] = "static_sticker";
    MessageSyncMessageTypes["ArEffect"] = "ar_effect";
    MessageSyncMessageTypes["SelfieSticker"] = "selfie_sticker";
})(MessageSyncMessageTypes = exports.MessageSyncMessageTypes || (exports.MessageSyncMessageTypes = {}));
